brave-core/src/main/java/com/github/kristofa/brave/InheritableServerClientAndLocalSpanState.java

```diff
@@ -0,0 +1,137 @@
+ * variables and provides local span inheritence from parent to children.
+ * <p>
+ * Important note: when using {@link InheritableServerClientAndLocalSpanState}, tracers must
+ * {@link LocalTracer#finishSpan() finish spans} or clear the local span at
+ * completion of the local trace span to avoid linking spans with incorrect
+ * parents and avoid leaking spans and associated memory.
+ */
+public final class InheritableServerClientAndLocalSpanState implements ServerClientAndLocalSpanState, InheritableState {
+
+    private final InheritableThreadLocal<ServerSpan> currentServerSpan =
+            new InheritableThreadLocal<ServerSpan>() {
+                @Override
+                protected ServerSpan initialValue() {
+                    return ServerSpan.EMPTY;
+                }
+            };
+
+    private final InheritableThreadLocal<Span> currentClientSpan = new InheritableThreadLocal<Span>();
+
+    private final InheritableThreadLocal<Deque<Span>> currentLocalSpan =
+            new InheritableThreadLocal<Deque<Span>>() {
+                @Override
+                protected Deque<Span> initialValue() {
+                    return new LinkedBlockingDeque<Span>();
+                }
+            };
+
+    private final Endpoint endpoint;
+
+    /**
+     * Constructor
+     *
+     * @param ip Int representation of ipv4 address.
+     * @param port port on which current process is listening.
+     * @param serviceName Name of the local service being traced. Should be lowercase and not <code>null</code> or empty.
+     */
+    public InheritableServerClientAndLocalSpanState(int ip, int port, String serviceName) {
+        Util.checkNotBlank(serviceName, "Service name must be specified.");
+        endpoint = Endpoint.create(serviceName, ip, port);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ServerSpan getCurrentServerSpan() {
+        return currentServerSpan.get();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setCurrentServerSpan(final ServerSpan span) {
+        if (span == null) {
+            currentServerSpan.remove();
+        } else {
+            currentServerSpan.set(span);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Endpoint endpoint() {
+        return endpoint;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Span getCurrentClientSpan() {
+        return currentClientSpan.get();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setCurrentClientSpan(final Span span) {
+        currentClientSpan.set(span);
+    }
+
+    @Override
+    public Boolean sample() {
+        return getCurrentServerSpan().getSample();
+    }
+
+    @Override
+    public Span getCurrentLocalSpan() {
+        return currentLocalSpan.get().peekFirst();
+    }
+
+    @Override
```

Comment:
here's where javadoc is helpful. you can mention the pop logic and anything to be careful of.